### PR TITLE
reduce unnecessary object update in workstatus_controller

### PR DESCRIPTION
**Pull request by MingZhang-YBPS**
**What type of PR is this?**

**What this PR does / why we need it**:
 The cluster object status change on member cluster triggers workstatus_controller. Then in function syncWorkStatus(), workstatus_controller checks if the object on Karmada control plane and member cluster are equal.
However the equivalent logic provided by function ObjectNeedsUpdate() always return true in case like below, which results in unneccessary ObjectWatcher.Update action from workstatus_controller.

Object in Karmada control plane like
```
metadata:
  creationTimestamp: "2023-07-10T05:58:42Z"
  generation: 1
  labels:
    clusterpropagationpolicy.karmada.io/name: xyz-propagation
```

Object in member cluster like 
```
metadata:
  labels:
    clusterpropagationpolicy.karmada.io/name: xyz-propagation
    clusterresourcebinding.karmada.io/key: 778949d7dc
    work.karmada.io/name: xyz-qii5l2vmpor7w7z06jnk-778949d7dc
    work.karmada.io/namespace: karmada-es-cluster170
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
In this PR, we solve the issue by ignoring all known Karmada labels of objects in both Karmada control plane and member cluster.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by chaunceyjiang** [comment]
> The cluster object status change on member cluster triggers workstatus_controller. Then in function syncWorkStatus(), workstatus_controller checks if the object on Karmada control plane and member cluster are equal.
However the equivalent logic provided by function ObjectNeedsUpdate() always return true in case like below, which results in unneccessary ObjectWatcher.Update action from workstatus_controller.

Similar to #3204.

**Comment by XiShanYongYe-Chang** [comment]
Hi @MingZhang-YBPS @chaunceyjiang, as I comment in https://github.com/karmada-io/karmada/pull/3792#pullrequestreview-1529531158, I think we can directly remove the labels compare when checking resource update. What do you think?

File: pkg/util/lifted/objectwatcher_test.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package lifted
 
 import (
 	"fmt"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
```
Comment (jwcesign) [inline pkg/util/lifted/objectwatcher_test.go:5]: import order

**Comment by XiShanYongYe-Chang** [comment]
Kindly ping @MingZhang-YBPS

**Comment by MingZhang-YBPS** [comment]
> Hi @MingZhang-YBPS @chaunceyjiang, as I comment in [#3792 (review)](https://github.com/karmada-io/karmada/pull/3792#pullrequestreview-1529531158), I think we can directly remove the labels compare when checking resource update. What do you think?

It sounds reasonable. 
And I am thinking if we should remove the entire code block https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L216-L235  

1. It is the responsible of execution controller rather than workstatus controller to sync object on member cluster with the one on Karmada control plane. workstatus controller shall only be used to sync object status on member cluster back to Karmada
2. And ObjectNeedsUpdate() is only used in workstatus controller

**Comment by MingZhang-YBPS** [comment]
> > Hi @MingZhang-YBPS @chaunceyjiang, as I comment in [#3792 (review)](https://github.com/karmada-io/karmada/pull/3792#pullrequestreview-1529531158), I think we can directly remove the labels compare when checking resource update. What do you think?
> 
> It sounds reasonable. And I am thinking if we should remove the entire code block https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L216-L235
> 
> 1. It is the responsible of execution controller rather than workstatus controller to sync object on member cluster with the one on Karmada control plane. workstatus controller shall only be used to sync object status on member cluster back to Karmada
> 2. And ObjectNeedsUpdate() is only used in workstatus controller

@RainbowMango How do you think about this?

**Comment by RainbowMango** [comment]
Will look at it later.

**Comment by XiShanYongYe-Chang** [comment]
Simply deleting this code block may not be enough.

Indeed, it does seem strange to perform resource synchronization in the workstatus controller.

For the execution controller, it can synchronize the changes in resource versions on the control plane to the member clusters. This process is top-down.

And for the workstatus controller, its synchronization operation is mainly to prevent users from modifying resources distributed by Karmada on the member cluster. This process is bottom-up.

Therefore, we still need to check whether the resources need to be synchronized in the workstatus controller.

Of course, this process can be improved, but I haven't thought of a better solution yet.

**Comment by MingZhang-YBPS** [comment]
> Simply deleting this code block may not be enough.
> 
> Indeed, it does seem strange to perform resource synchronization in the workstatus controller.
> 
> For the execution controller, it can synchronize the changes in resource versions on the control plane to the member clusters. This process is top-down.
> 
> And for the workstatus controller, its synchronization operation is mainly to prevent users from modifying resources distributed by Karmada on the member cluster. This process is bottom-up.
> 
> Therefore, we still need to check whether the resources need to be synchronized in the workstatus controller.
> 
> Of course, this process can be improved, but I haven't thought of a better solution yet.

OK. I got the background. So it seems ok to remove the label comparsion when checking resource update (default retain webhook respects label change on member cluster anyway)

**Comment by XiShanYongYe-Chang** [comment]
Okay, we can do it by removing the comparison of the label and then adding some comments to explain the reasons. Thank you~

**Comment by RainbowMango** [comment]
Hi @MingZhang-YBPS I tried to reproduce this issue on my side([by this sample](https://github.com/karmada-io/karmada/tree/master/samples/nginx)). I found that the `labels` on the Karmada control are same with the member cluster:

Here is the manifest in the `Work`:
```yaml
spec:
  workload:
    manifests:
    - apiVersion: apps/v1
      kind: Deployment
      metadata:
        annotations:
          kubectl.kubernetes.io/last-applied-configuration: |
            {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"nginx","namespace":"default"},"spec":{"replicas":2,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata}
          resourcebinding.karmada.io/name: nginx-deployment
          resourcebinding.karmada.io/namespace: default
          resourcetemplate.karmada.io/managed-annotations: kubectl.kubernetes.io/last-applied-configuration,resourcebinding.karmada.io/name,resourcebinding.karmada.io/namespace,resourcetemplate.karmada.io/managed-annotations,resourced
          resourcetemplate.karmada.io/managed-labels: app,karmada.io/managed,propagationpolicy.karmada.io/name,propagationpolicy.karmada.io/namespace,resourcebinding.karmada.io/key,work.karmada.io/name,work.karmada.io/namespace
          resourcetemplate.karmada.io/uid: 2f77d290-2f23-443a-a4fd-ffcf4e7520ed
        labels:
          app: nginx
          karmada.io/managed: "true"
          propagationpolicy.karmada.io/name: nginx-propagation
          propagationpolicy.karmada.io/namespace: default
          resourcebinding.karmada.io/key: 5bdc5c4dcf
          work.karmada.io/name: nginx-687f7fb96f
          work.karmada.io/namespace: karmada-es-member1
```

Here is the yamls in member cluster:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"nginx","namespace":"default"},"spec":{"replicas":2,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"la}
    resourcebinding.karmada.io/name: nginx-deployment
    resourcebinding.karmada.io/namespace: default
    resourcetemplate.karmada.io/managed-annotations: kubectl.kubernetes.io/last-applied-configuration,resourcebinding.karmada.io/name,resourcebinding.karmada.io/namespace,resourcetemplate.karmada.io/managed-annotations,resourcetemplad
    resourcetemplate.karmada.io/managed-labels: app,karmada.io/managed,propagationpolicy.karmada.io/name,propagationpolicy.karmada.io/namespace,resourcebinding.karmada.io/key,work.karmada.io/name,work.karmada.io/namespace
    resourcetemplate.karmada.io/uid: 2f77d290-2f23-443a-a4fd-ffcf4e7520ed
  creationTimestamp: "2023-07-17T06:45:53Z"
  generation: 1
  labels:
    app: nginx
    karmada.io/managed: "true"
    propagationpolicy.karmada.io/name: nginx-propagation
    propagationpolicy.karmada.io/namespace: default
    resourcebinding.karmada.io/key: 5bdc5c4dcf
    work.karmada.io/name: nginx-687f7fb96f
    work.karmada.io/namespace: karmada-es-member1
  name: nginx
```

They both have `app`/`karmada.io/managed`/`propagationpolicy.karmada.io/name`/... labels.

In the PR description, you mentioned that 
```
Object in Karmada control plane like

metadata:
  creationTimestamp: "2023-07-10T05:58:42Z"
  generation: 1
  labels:
    clusterpropagationpolicy.karmada.io/name: xyz-propagation
```
I guess this piece of yaml comes from the `resource template` instead of `Work`, Could you please confirm that?
If yes, could you please describe how to reproduce it? I think we need to figure out the root case and make solutions based on that.

**Comment by MingZhang-YBPS** [comment]
> Hi @MingZhang-YBPS I tried to reproduce this issue on my side([by this sample](https://github.com/karmada-io/karmada/tree/master/samples/nginx)). I found that the `labels` on the Karmada control are same with the member cluster:
> 
> Here is the manifest in the `Work`:
> 
> ```yaml
> spec:
>   workload:
>     manifests:
>     - apiVersion: apps/v1
>       kind: Deployment
>       metadata:
>         annotations:
>           kubectl.kubernetes.io/last-applied-configuration: |
>             {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"nginx","namespace":"default"},"spec":{"replicas":2,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata}
>           resourcebinding.karmada.io/name: nginx-deployment
>           resourcebinding.karmada.io/namespace: default
>           resourcetemplate.karmada.io/managed-annotations: kubectl.kubernetes.io/last-applied-configuration,resourcebinding.karmada.io/name,resourcebinding.karmada.io/namespace,resourcetemplate.karmada.io/managed-annotations,resourced
>           resourcetemplate.karmada.io/managed-labels: app,karmada.io/managed,propagationpolicy.karmada.io/name,propagationpolicy.karmada.io/namespace,resourcebinding.karmada.io/key,work.karmada.io/name,work.karmada.io/namespace
>           resourcetemplate.karmada.io/uid: 2f77d290-2f23-443a-a4fd-ffcf4e7520ed
>         labels:
>           app: nginx
>           karmada.io/managed: "true"
>           propagationpolicy.karmada.io/name: nginx-propagation
>           propagationpolicy.karmada.io/namespace: default
>           resourcebinding.karmada.io/key: 5bdc5c4dcf
>           work.karmada.io/name: nginx-687f7fb96f
>           work.karmada.io/namespace: karmada-es-member1
> ```
> 
> Here is the yamls in member cluster:
> 
> ```yaml
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   annotations:
>     deployment.kubernetes.io/revision: "1"
>     kubectl.kubernetes.io/last-applied-configuration: |
>       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"nginx","namespace":"default"},"spec":{"replicas":2,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"la}
>     resourcebinding.karmada.io/name: nginx-deployment
>     resourcebinding.karmada.io/namespace: default
>     resourcetemplate.karmada.io/managed-annotations: kubectl.kubernetes.io/last-applied-configuration,resourcebinding.karmada.io/name,resourcebinding.karmada.io/namespace,resourcetemplate.karmada.io/managed-annotations,resourcetemplad
>     resourcetemplate.karmada.io/managed-labels: app,karmada.io/managed,propagationpolicy.karmada.io/name,propagationpolicy.karmada.io/namespace,resourcebinding.karmada.io/key,work.karmada.io/name,work.karmada.io/namespace
>     resourcetemplate.karmada.io/uid: 2f77d290-2f23-443a-a4fd-ffcf4e7520ed
>   creationTimestamp: "2023-07-17T06:45:53Z"
>   generation: 1
>   labels:
>     app: nginx
>     karmada.io/managed: "true"
>     propagationpolicy.karmada.io/name: nginx-propagation
>     propagationpolicy.karmada.io/namespace: default
>     resourcebinding.karmada.io/key: 5bdc5c4dcf
>     work.karmada.io/name: nginx-687f7fb96f
>     work.karmada.io/namespace: karmada-es-member1
>   name: nginx
> ```
> 
> They both have `app`/`karmada.io/managed`/`propagationpolicy.karmada.io/name`/... labels.
> 
> In the PR description, you mentioned that
> 
> ```
> Object in Karmada control plane like
> 
> metadata:
>   creationTimestamp: "2023-07-10T05:58:42Z"
>   generation: 1
>   labels:
>     clusterpropagationpolicy.karmada.io/name: xyz-propagation
> ```
> 
> I guess this piece of yaml comes from the `resource template` instead of `Work`, Could you please confirm that? If yes, could you please describe how to reproduce it? I think we need to figure out the root cause and make solutions based on that.

Hi @RainbowMango 
You are right, I checked the log again, the object update happened in workstatus controller continuously, that is  because the admission on member cluster reject the update. Then async worker in workstatus controller enters into retrying.
So the most important problem here is how to notify user that "something is out-of-sync" (see https://github.com/karmada-io/karmada/issues/3572#issuecomment-1634249760). And once user solve it, the workstatus controller will recover automatically.
Let's focus on https://github.com/karmada-io/karmada/issues/3572#issuecomment-1634249760

**Pull request closed without merging**